src/test/java/com/google/sps/CommentServiceTest.java

```diff
@@ -0,0 +1,87 @@
+package com.google.sps;
+
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import com.google.sps.servlets.CommentService;
+import java.util.List;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+
+public final class CommentServiceTest {
+
+  CommentService commentService = new CommentService();
+
+  /**
+   * A list of Strings should be returned for any public video that has comments. 
+   */
+  @Test
+  public void publicVideoReturnsCommentsList() throws GeneralSecurityException, IOException {
+    String videoId = "v3whnmU619I"; // Public video that has some comments.
+    List<String> actual = commentService.getCommentsFromId(videoId);
+
+    Assert.assertFalse(actual.isEmpty());
+  }
+
+  /**
+   * Unlisted videos should be accessible to the API if the video id is inputted. List of comments
+   * should be returned for unlisted videos just like they would for a public video. 
+   */
+  @Test
+  public void unlistedVideoReturnsCommentsList() throws GeneralSecurityException, IOException {
+    String videoId = "9BUFUUhfXSg"; // Unlisted test video that has 2 comments. 
+    List<String> actual = commentService.getCommentsFromId(videoId);
+
+    Assert.assertFalse(actual.isEmpty());
+  }
+
+  /**
+   * Private videos should not be accessible with the YouTube API. In this case, a 
+   * GoogleJsonResponseException should be caught and null should be returned. 
+   */
+  @Test
+  public void privateVideoReturnsNull() throws GeneralSecurityException, IOException {
+    String videoId = "qYnSnsysgVk";
+    List<String> actual = commentService.getCommentsFromId(videoId);
+
+    Assert.assertEquals(null, actual);
+  }
+  
+  /**
+   * For videos that have disabled commenting, a GoogleJsonResponseException should be caught and
+   * null should be returned. 
+   */
+  @Test
+  public void disabledCommentsReturnsNull() throws GeneralSecurityException, IOException {
+    String videoId = "NEXFoP0JurI";
+    List<String> actual = commentService.getCommentsFromId(videoId);
+
+    Assert.assertEquals(null, actual);
+  }
+
+  /**
+   * Videos with no comments should return null to avoid lowering the average sentiment score. 
+   */
+  @Test
+  public void noCommentsReturnsNull() throws GeneralSecurityException, IOException {
+    String videoId = "ERL0EDZT1kE";
+    List<String> actual = commentService.getCommentsFromId(videoId);
+
+    Assert.assertEquals(null, actual);
+  }
+
+  /**
+   * Invalid IDs that don't refer to a video should return null. 
+   */
```

Comment:
should we test this case in the backend or in the frontend before we even send the data to the backend? or both??

Reply:
I think there are two types of bad ids:

1. Those that are apparently illegal without calling the api, like "", or "123", or contains illegal characters like "abc//abc", we can certainly check these type of error in the front end. 
2. Those that appear legal but aren't, like replacing a few characters in a proper video id. We have to call the api in this case.

My suggestion is that we can have a `isValidId` checker class in the back end before making other api calls. It will certainly be slower than if we check it in the front end (in the case of type 1 illegal id), but I think it is a good idea to  check the validity in one place.

This would benefit the `Caption` class too :)

Reply:
I agree with checking valid IDs in the front end, one thing I'll mention is there doesn't seem to be any pattern to the IDs so I don't know how we can define something like "123" as "apparently illegal".

Reply:
If someone else can do the job for you and return to you a nice error then you shouldn't do it yourself :). Haven't read through your implementation yet but if you call the API with a bad video id and it tells you it is a bad id, just catch it and tell the client that they need to input a valid id.